Add ability for host to disable spawns in game lobby

ocdi opened the pull request:
This PR introduces the ability for a host to mark closed slots as occupying certain spawn locations.

This shows a multiplayer game which is syncing the closed slot state.

![image](https://user-images.githubusercontent.com/2775804/87784751-3b9e0b00-c87a-11ea-95c2-b4441bb9cbd0.png)

I bumped the LangVersion to 6 to enable the null propagation operator for one line of code and can be altered to remove the C#6 feature, but having said that is there a reason this project is targeting version 5 still?

Closes #18298.

ocdi committed:
Add ability to track closed slots and update dropdown to incorporate this.

ocdi committed:
Handle closed slot orders and update the UI to incorporate this

ocdi committed:
Add SPAWN_DROPDOWN to lobby-players UI definitions

ocdi committed:
Allow the SPAWN_DROPDOWN to be null if not defined in the template.

teinarss commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs:
```diff
@@ -231,9 +231,17 @@ public SlotDropDownOption(string title, string order, Func<bool> selected)
 				.ToDictionary(c => spawns[c.SpawnPoint - 1], c => new SpawnOccupant(c));
+
+			foreach (var closedSlot in lobbyInfo.Slots.Values.Where(s => s.Closed && s.ClosedSpawnPoint - 1 >= 0 && s.ClosedSpawnPoint - 1 < spawns.Length))
+			{
+				occupants[spawns[closedSlot.ClosedSpawnPoint - 1]] = new SpawnOccupant(closedSlot);
+			}
```
No need for curly bracers when its only a single line.

teinarss commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs:
```diff
@@ -413,6 +430,15 @@ public static void SetupNameWidget(Widget parent, Session.Slot s, Session.Client
 			slot.GetText = () => truncated.Update(c != null ? c.Name : s.Closed ? "Closed" : "Open");
 			slot.OnMouseDown = _ => ShowSlotDropDown(slot, s, c, orderManager, map);
 
+			if (s.Closed)
+			{
+				SetupEditableSpawnWidget(parent, s, null, orderManager, map);
+			}
+			else
+			{
+				HideChildWidget(parent, "SPAWN_DROPDOWN");
+			}
+
```
Same here.

teinarss reviewed (changes requested):
Thanks for the PR! Just some small style problems with the code.

pchote commented:
I haven't looked at this properly, but there is one potential issue that needs to be considered: most maps define 1:1 spawns to slots, but this is a convention rather than a requirement and I believe that there are some community maps that deliberately define more spawns than player slots.

kyrreso commented:
Nice!
 #18298 

Next step could also be to have a X on the spot in the minimap, but that can be added later...

ocdi committed:
Address PR feedback

ocdi committed:
Exclude closed spawn locations from random allocation

ocdi commented:
I updated the PR based on your comments. I was also thinking that the random allocation would not be affected by my change so also added another change that factors them in.

ocdi commented:
@pchote This is certainly a limitation of this approach. It makes sense from a UI perspective to implement it the way I have - I can't think how to model this any other way without redesigning the UI. Although I guess being able to block some locations is better than not right? :-)

teinarss commented:
Regression: Not possible to change spawn by clicking on the map when a spawn location is already chosen. It works if clearing the spawn location first.

ocdi commented:
@teinarss Is this where no spawn locations you can click on the map and the spawns will be allocated for the first unallocated for existing players / bots. If you chose a slot as closed and click, it doesn't continue to select beyond that? 

Is the ask to make the new behaviour match the old such that clicking should enable selection for the closed slots also? I am assuming so and will work on that next. 👍

abcdefg30 commented:
Would it make sense to allow left clicking to assign spawns, and right click to disable one? (With a visible indicator, maybe crossing the point out.) I'm not sure about assigning spawns to closed slots, since it looks a bit odd imo. (But don't really have a better idea right now other than doing ^ which can be done additionally as well.)

ocdi committed:
Allow RMB to select closed slot locations and clear spawn point when reopening a slot

ocdi committed:
Only consider closed slots for map selection and ensure closed slots can't be chosen via the map

ocdi commented:
OK based on @teinarss and @abcdefg30 's feedback I have updated the PR to incorporate both of those suggestions. I compared before/after the behaviour of the map based on the logic.

This shows using left mouse button to select all the slots as a game host
![lefts](https://user-images.githubusercontent.com/2775804/87864208-fdebd000-c9a8-11ea-85ea-b745e901c80a.gif)

It supports using the RMB to select only closed slots.
![rights](https://user-images.githubusercontent.com/2775804/87864211-03e1b100-c9a9-11ea-92bd-170edd0559ab.gif)

Once slots are full clicking away from a slot unselects the first position - for a host this is the first player on the LMB and the first closed slot for a RMB.

I am going to see if I can draw an X next.

ocdi committed:
Allow non-host to continue cloking to select map positions.

ocdi commented:
Actually further to above I discovered / understood the issue actually raised by @teinarss - the client local index was used in the slot selection, so for non-hosts you could continue clicking and it would update the local selection. I do think the other behaviour I introduced is more beneficial though. I have learned a lot in understanding how OpenRA works through doing this PR.

ocdi committed:
Draw an X over the closed slots

ocdi commented:
Last commit of the day is fairly self explanatory. It is very crude, but I am drawing an X using BigBold font over the circle. I am unsure if this should be on top of the spawn letter or as I did it in this screenshot:

![image](https://user-images.githubusercontent.com/2775804/87864840-a4879f00-c9b0-11ea-9ee7-55836196e806.png)

I also just observed that my last commit actually works through all selections (including slots) when clicking on a map away from a slot, which is a nice surprise, and that honours the LMB (yellow) vs RMB (blue).

![2020-07-19-11-15-45](https://user-images.githubusercontent.com/2775804/87864886-53c47600-c9b1-11ea-9f74-990a76a7d1d2.gif)

ocdi commented:
So I was thinking and re-reading through some of the comments and I just realised I had blinders on and had not actually addressed some of your concerns. I took what I learned and I have created a new PR that I think addresses the asks better. 
#18425 

I actually prefer my new PR over this one so will close this in favour of my newer version.
